EIPS/eip-7748.md

```diff
@@ -0,0 +1,292 @@
+status: Draft
+type: Standards Track
+category: Core
+created: 2024-07-23
+requires: 7612
+---
+
+## Abstract
+
+This EIP proposes a procedure to convert, on each block, a fixed number of key-values from the existing  Merkle Patricia Tree (MPT) to the Verkle Tree (VKT).
+
+## Motivation
+
+The accounts state is too large to wait for transactions to organically move all of them to the VKT through the Overlay Tree. Thus, we need a strategy to convert all the state within a reasonable time. The state conversion completion allows removing the Overlay Tree abstraction introduced in [EIP-7612](./eip-7612.md) and use directly the VKT for all state access.
+
+## Specification
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119 and RFC 8174.
+
+### Constants
+
+| Parameter                    | value | Description                                                    |
+| ---------------------------- | ----- | -------------------------------------------------------------- |
+| `CONVERSION_START_TIMESTAMP` | `TBD` | Timestamp at which the conversion starts.                      |
+| `CONVERSION_STRIDE`          | `TBD` | Maximum number of _conversion units_ to be converted per block |
+
+A _conversion unit_ is:
+
+- A contract storage slot.
+- A contract code. (i.e. all the code is a single _conversion unit_)
+- An account data. (e.g. balance, nonce, code-hash)
+
+### Changes to the execution spec
+
+Include the following code in the existing `apply_body(...)` function:
+
+```python
+def apply_body(state: State, ...) -> Tuple[Uint, Root, Root, Bloom, State, Root]:
+    ...
+    # <new_code>
+    if block_time >= CONVERSION_START_TIMESTAMP and not state._conversion_finished:
+        block_state_conversion(state, CONVERSION_STRIDE)
+    # </new_code>
+    
+    for i, tx in enumerate(map(decode_transaction, transactions)):
+        ...
+    ...
+```
+
+Before executing txs, it calls `block_state_conversion(...)` (described below) which performs a state conversion step for this block.
+
+In `state.py`, add the following code:
+
+```python
+@dataclass
+class StoragePhase:
+    """
+    The account next conversion step continues converting the 
+    storage-slot with key greater or equal next_key. 
+    If there isn't such storage-slot, the account must move to
+    BasicDataPhase.
+    """
+    next_key: Bytes   
+   
+@dataclass
+class BasicDataPhase:
+    """
+    The account next conversion step continues migrating the account 
+    code (if any) and basic data. After processing, the account must 
+    move to the next account in the trie (or finish if it was the 
+    last one).
+    """
+    pass
+
+@dataclass
+class CurrentConvertingAccount:
+    """
+    Contains the state conversion next step.
+    """
+    address: Address
+    phase : StoragePhase | BasicDataPhase
```

Comment:
Later in the `state_convert` method there is a reference to a `CodePhase`. Should this also be added?

Reply:
Oh sorry, this is a mistake. On a previous version, I was moving code-chunks as conversion units. But later we simplified to move all the code in one go.

I'll open a PR to fix this soon (and we can decide on the renaming above in the PR too)

Thanks!